### PR TITLE
Recurrent tasks

### LINKED ISSUE #86: new extensions - Recurrence
It would be great to have the recurrence extension like simpletask for android!
(https://github.com/bram85/todo.txt-tools/wiki/Recurrence)

### PR DESCRIPTION
**Pull request by topas-rec**
Dear folks,

I was working on recurring tasks. This should fix #86. This is the initial result. I am not finished.
I reached a point where I need your ideas:

As you seen in the todo.txt file in this branch when tasks have creation dates and are recurring the creation date is recreated in the recurring task when the old task is completed. I think it happens during construction of the recurring task. I would like to get your ideas how and where this creation date recreation can be avoided. (removeCreationDate function? or if else in Task constructor?)

(Disabled editing - would like to learn and do it myself)
Yes - my commits are a mess. I didn't understand the synchronization with github properly. There are also very much changed files. Do I have to rebase the branch? Is that what causes so many file changes? I changed only three or four files.

Thanks for you help!

### CODE REVIEW

**Comment by oroulet** [comment]
Try to do 'git pull -r httpsxxxxxmainrep master'

To clean it even further use git rebase -i and squash all commits

**Comment by oroulet** [comment]
@topas-rec Can you try to fix that one? or do you have issues?

**Comment by topas-rec** [comment]
Thanks for your help! I missed that you already commented.
I would like to try it by myself. (I cannot work on it right now cause I was send to US from work)

**Comment by topas-rec** [comment]
I'm back on Wednesday.

**Comment by topas-rec** [comment]
Hey @oroulet I cannot get the thing to work. I tried to rebase by branch onto the master of QTodoTxt. Can you help me with further questions or instructions?
I created the branch recurring_tasks and then fetched and merged changes from upstream by committing them into my branch and pushing them into the origin remote branch.

**Comment by topas-rec** [comment]
> Try to do 'git pull -r https://github.com/QTodoTxt/QTodoTxt.git master'
To clean it even further use git rebase -i and squash all commits

Do I need to commit and push after that? (Haven't done that last time)

**Comment by oroulet** [comment]
yes do a 'git push --force'

**Comment by topas-rec** [comment]
OK  @oroulet I was able to rebase it (a bit).
I have to get it even more cleaner but now I have an idea on how to do it.

**Comment by topas-rec** [comment]
As mentioned above:
> As you seen in the todo.txt file in this branch when tasks have creation dates and are recurring the creation date is recreated in the recurring task when the old task is completed. I think it happens during construction of the recurring task. I would like to get your ideas how and where this creation date recreation can be avoided. (removeCreationDate function? or if else in Task constructor?)

I still have a question.
Furthermore I want to get the functionality working that makes a task recur from the original due date. (Exactly in the same way simpletask does it with the "plus" sign in front of the recurring time tag)

**Comment by topas-rec** [comment]
Cleaned it further as much as possible with the commands you ( @oroulet ) provided me. Is there anything else I could do?
My egit shows changes in three files when I compare local branch recurring_tasks with upstream branch master. Why does github shows 9 changed files?

**Comment by topas-rec** [comment]
OK I made it! Awesome. Aborting an interactive rebase showed me suspicious changes in egit. (For the record) Commiting (forced) then helped.

**Comment by topas-rec** [comment]
Everyone, please answer the questions posed above if interested.

**Comment by oroulet** [comment]
Look at my comment in the other PR and to clean this one you to need to use "git rebase -i" and squash all commits

**Comment by topas-rec** [comment]
I'll try to further clean this up.

**Comment by topas-rec** [comment]
Can someone please test this? It should work like this: https://github.com/mpcjanssen/simpletask-android/blob/master/src/main/assets/index.en.md

**Comment by topas-rec** [comment]
I just found an issue: 'b' is not implemented. So repeating tasks by workday wouldn't work. (I'll work on this as soon as possible)

**Comment by topas-rec** [comment]
I cannot get this done. I'm using eclipse and I think it has a bug when I want to rebase. The implementation is done in my eyes, but I cannot contribute this.

**Comment by oroulet** [comment]
I will fix it when I get two minutes

**Comment by oroulet** [comment]
Get the code working and clean and I will do the git stuff

**Comment by topas-rec** [comment]
For the record: How to interactively rebase in eclipse: (squash commits): https://www.youtube.com/watch?v=2nv9O-LaC_0

**Comment by topas-rec** [comment]
Can someone tell me if and how I can execute completeTask(self, task) from task_list_controller in a pytest? Do I need to instanciate the whole app?

**Comment by topas-rec** [comment]
I would do this last unit test. I think I am done when it is finished

File: qtodotxt/lib/tasklib.py [hunk 1]
```diff
@@ -44,6 +44,11 @@ def _reset(self):
         self.due_error = ""
         self.threshold = None
         self.keywords = {}
+        self.recMode = None
+        self.recModeCompl = 0   # Original due date mode: Task recurs from original due date
+        self.recModeOrDue = 1   # Completion date mode: Task recurs from completion date
```
Comment (oroulet) [inline qtodotxt/lib/tasklib.py]: As Far as I understand you should create an Enum called RecursiveMode. Not 2 class variables

File: qtodotxt/lib/tasklib.py [hunk 2]
```diff
@@ -44,6 +44,11 @@ def _reset(self):
...
+        self.recIncr = None
+        self.recInterv = None
```
Comment (oroulet) [inline qtodotxt/lib/tasklib.py]: Maybe you should have 2 class called Recurtion with 3 members: Mode (an Enum), Increment and Interval (Do not use so much shortcuts, it make it very hard to understand)
The would recMode be None if no recursion and a Recurtion instance if task is recuring. There are other possibilities but we definitely do not need 5 intance variables

File: qtodotxt/ui/controllers/tasks_list_controller.py [hunk 3]
```diff
@@ -105,6 +107,11 @@ def _initIncreasePrioritySelectedTasksAction(self):
 
     def completeTask(self, task):
         if not task.is_complete:
+            # Check if task is recurrent
+            # TODO: Many variables checked here - does it work smarter, too?
+            if task.recIncr is not None and task.recInterv is not None \
+                    and task.recMode is not None and task.due is not None:
+                self._recurTask(task)
```
Comment (oroulet) [inline qtodotxt/ui/controllers/tasks_list_controller.py]: that will simplify that test a lot

File: qtodotxt/ui/controllers/tasks_list_controller.py [hunk 4]
```diff
@@ -184,16 +248,23 @@ def _addCreationDate(self, text):
...
-        (text, ok) = self._task_editor_service.createTask()
+    def createTask(self, arg_text=None):
+        if not arg_text:
+            (text, ok) = self._task_editor_service.createTask()
+        else:
+            text = arg_text
+            ok = True
         if ok and text:
             if int(QtCore.QSettings().value("add_created_date", 0)):
+                # TODO: Remove existing creation date first
```
Comment (oroulet) [inline qtodotxt/ui/controllers/tasks_list_controller.py]: why do you have a TODO line here?

File: qtodotxt/ui/controllers/tasks_list_controller.py [hunk 5]
```diff
@@ -176,6 +229,17 @@ def _reselect(self, tasks):
     def _sortTasks(self, tasks):
         tasks.sort(reverse=True)
 
+    def _removeCreationDate(self, text):
+        # Find the date that has no keyword in advance just whitespaces
+        # TODO: This removes the priority of recurrent tasks
```
Comment (oroulet) [inline qtodotxt/ui/controllers/tasks_list_controller.py]: why a todo here too? It looks like it is already implemented.

**Comment by topas-rec** [comment]
Thank you. I'll try to fox these points.

File: qtodotxt/lib/tasklib.py (hunk 1 above)
Comment (topas-rec) [inline qtodotxt/lib/tasklib.py]: Yes. I remember I didn't like these values as class variables, too.

Comment (topas-rec) [inline qtodotxt/lib/tasklib.py]: Working on that first

File: qtodotxt/lib/tasklib.py (hunk 2 above)
Comment (topas-rec) [inline qtodotxt/lib/tasklib.py]: Accepted - working on it.

File: qtodotxt/ui/controllers/tasks_list_controller.py (hunk 3 above)
Comment (topas-rec) [inline qtodotxt/ui/controllers/tasks_list_controller.py]: Yes, but I can make this test also simpler if I just would check the variable `task.recMode is not None`
I've redesigned the class structure(s) to your suggestions. Thank you.

File: qtodotxt/ui/controllers/tasks_list_controller.py (hunk 4 above)
Comment (topas-rec) [inline qtodotxt/ui/controllers/tasks_list_controller.py]: Removed

File: qtodotxt/ui/controllers/tasks_list_controller.py (hunk 5 above)
Comment (topas-rec) [inline qtodotxt/ui/controllers/tasks_list_controller.py]: Yes, is already done. Removed

File: qtodotxt/lib/tasklib.py [hunk 6]
```diff
@@ -2,10 +2,27 @@
 import re
 
 from PyQt5 import QtCore
-
+from enum import Enum
 from qtodotxt.lib.task_htmlizer import TaskHtmlizer
 
 
+class recursiveModes(Enum):
+    init = 0     # Initial recursive mode
```
Comment (oroulet) [inline qtodotxt/lib/tasklib.py]: This is probably not necessary. but call it None if you really need it.

File: qtodotxt/lib/tasklib.py [hunk 7]
```diff
@@ -2,10 +2,27 @@
...
+    mode = recursiveModes.init
+    increment = None
+    interval = None
+
+    def __init__(self, arg_mode, arg_increment, arg_interval):
+        self.mode = arg_mode
+        self.increment = arg_increment
+        self.interval = arg_interval
+
+
```
Comment (oroulet) [inline qtodotxt/lib/tasklib.py]: Good

File: qtodotxt/lib/tasklib.py (hunk 6 above)
Comment (oroulet) [inline qtodotxt/lib/tasklib.py]: btw, maybe it is cleaner to use RecursiveMode.None instead og setting class member to None. I let you decide.

Comment (oroulet) [inline qtodotxt/lib/tasklib.py]: should be called RecursiveMode not RecursiveModeS

File: qtodotxt/test/test_tasks.py [hunk 8]
```diff
@@ -182,3 +182,46 @@ def test_creation_date(self):
...
+        task = Task('(C) do something due:2016-09-05 rec:3m')
+        self.assertIsNotNone(task.recursion)
+        self.assertTrue(task.recursion.mode == recursiveModes.completitionDate)
+        self.assertTrue(task.recursion.increment == str(3))
+        self.assertTrue(task.recursion.interval == 'm')
+
+    def test_recurring_task_input_years(self):
+        task = Task('(C) do something due:2016-09-05 rec:+1y')
+        self.assertIsNotNone(task.recursion)
+        self.assertTrue(task.recursion.mode == recursiveModes.originalDueDate)
```
Comment (oroulet) [inline qtodotxt/test/test_tasks.py]: looks good

File: qtodotxt/ui/controllers/tasks_list_controller.py [hunk 9]
```diff
@@ -105,6 +108,10 @@ def _initIncreasePrioritySelectedTasksAction(self):
 
     def completeTask(self, task):
         if not task.is_complete:
+            # Check if task is recurrent
+            # TODO: Many variables checked here - does it work smarter, too?
```
Comment (oroulet) [inline qtodotxt/ui/controllers/tasks_list_controller.py]: remove TODO

File: qtodotxt/lib/tasklib.py (hunk 6 above)
Comment (topas-rec) [inline qtodotxt/lib/tasklib.py]: Init is not really needed. None is fine.

**Comment by oroulet** [comment]
I let you fix it then. Then I will merge

File: qtodotxt/lib/tasklib.py (hunk 6 above)
Comment (topas-rec) [inline qtodotxt/lib/tasklib.py]: > btw, maybe it is cleaner to use RecursiveMode.None instead og setting class member to None. I let you decide.

Sorry, don't know how to use it. Can I alternatively initialize something in this way?

Comment (topas-rec) [inline qtodotxt/lib/tasklib.py]: > should be called RecursiveMode not RecursiveModeS
`class recursiveModes(Enum):
    completitionDate = 0    # Original due date mode: Task recurs from original due date
    originalDueDate = 1    # Completion date mode: Task recurs from completion date`
should list all possible modes. Why not calling it recursiveModes?
The mode itself is set in 
`class recursion:
    mode = None
    increment = None
    interval = None`
and the used variable which only holds THE current mode is called `mode`.

File: qtodotxt/test/test_tasks.py (hunk 8 above)
Comment (topas-rec) [inline qtodotxt/test/test_tasks.py]: I would like to create a task which is recurrent and then complete it and test if another one with the correct due date is there. I cannot get this done. Do you have an idea?

**Comment by oroulet** [comment]
Enum are singular by convention,....

File: qtodotxt/ui/controllers/tasks_list_controller.py (hunk 9 above)
Comment (topas-rec) [inline qtodotxt/ui/controllers/tasks_list_controller.py]: Done

**Comment by topas-rec** [comment]
I see. I didn't know.

**Comment by oroulet** [comment]
merging that one so maybe other people test. You can always improve later.

**Comment by topas-rec** [comment]
Thanks. Should I create an issue for the missing test?

**Comment by topas-rec** [comment]
The readme file is outdated from now. Should I also create an issue for that or just a pull request?

**Comment by oroulet** [comment]
Pr is always better than bug report ;-)

**Pull request merged**